src/org/rascalmpl/library/util/Eval.java

```diff
@@ -48,169 +61,278 @@
 				}
-				timer.cancel();
-				if (timer.hasExpired()) {
-					throw RuntimeExceptionFactory.timeout(null, null);
+
+				if (URIUtil.getParentLocation(current).equals(current)) {
+					// we went all the way up to the root
+					return reg.isDirectory(member) ? member : URIUtil.getParentLocation(member);
 				}
 				
-				if (expected != null) {
-					Type typ = tr.valueToType((IConstructor) expected);
-					if (!result.getStaticType().isSubtypeOf(typ)) {
-						throw new UnexpectedType(typ, result.getStaticType(), URIUtil.rootLocation("eval"));
-					}
-				}
-				return result;
+				current = URIUtil.getParentLocation(current);
 			}
+
+			return current;
 		}
-		catch (ParseError e) {
-			if (forRascal) 
-				throw RuntimeExceptionFactory.parseError(e.getLocation(), null, null);
-			throw e;
-		}
-		catch (StaticError e) {
-			if (forRascal) {
-			  throw new Throw(values.constructor(Exception_StaticError, values.string(e.getMessage()), e.getLocation()), (ISourceLocation) null, null);
+		
+		private String javaCompilerPathAsString(IList javaCompilerPath) {
+			StringBuilder b = new StringBuilder();
+
+			for (IValue elem : javaCompilerPath) {
+				ISourceLocation loc = (ISourceLocation) elem;
+
+				if (b.length() != 0) {
+					b.append(File.pathSeparatorChar);
+				}
+
+				// this is the precondition
+				assert loc.getScheme().equals("file");
+
+				// this is robustness in case of experimentation in pom.xml
+				if ("file".equals(loc.getScheme())) {
+					b.append(Paths.get(loc.getURI()).toAbsolutePath().toString());
+				}
 			}
-			throw e;
-		} 
-		catch (URISyntaxException e) {
-			// this should never happen
-			if (forRascal)
-				throw RuntimeExceptionFactory.illegalArgument(commands, null, null);
-			throw new RuntimeException(e.getMessage(), e);
+
+			return b.toString();
 		}
-		finally {
-			evaluator.getHeap().removeModule(env);
-			evaluator.setCurrentEnvt(old);
+	
+		public void reset() {
+			eval.getCurrentModuleEnvironment().reset();
+			eval.getHeap().clear();
+		}
+
+		public Result<IValue> eval(IRascalMonitor monitor, String line) throws InterruptedException, IOException {
+			return eval.eval(monitor, line, IRascalValueFactory.getInstance().sourceLocation(URIUtil.assumeCorrect("eval", "", "", "command=" + line)));
 		}
-		
-		if (forRascal) 
-			throw RuntimeExceptionFactory.illegalArgument(commands, null, null);
-		
-		throw new IllegalArgumentException();
 	}
-	
 
-	public static class Timer extends Thread {
+	/**
+	 * This Thread class counts 10 steps towards overflowing the timeout,
+	 * and then stops while informing the given Evaluator object to interrupt
+	 * its running program.
+	 */
+	public static class EvaluatorInterruptTimer extends Thread {
+		private final Evaluator eval;
 		private final int timeout;
 		private int elapsed;
 		private final int sample;
 		private volatile boolean running;
-
-		public Timer(int timeout) {
+		
+		public EvaluatorInterruptTimer(Evaluator eval, int timeout) {
```

Comment:
if we keep it like this, lets call `setDaemon(true)` to make sure they won't keep the jvm alive at close time.

Reply:
check